### PR TITLE
UI Link

### PR DESCRIPTION
**Pull request by alex40724**

### CODE REVIEW

**Review by klees — CHANGES_REQUESTED** [review]
Hi @alex40724,
I'm not happy with this. My reasons are threefold:

* I'm not sure, we really need this component in this form. As far as I understand, atm you need this for the dropdown. Why don't you just use a button? Is it just a matter of presentation?
* The naming to me is wrong. A [link](https://de.wikipedia.org/wiki/Hyperlink) references another resource in hypertext content. As the description states, this will not always be the case with this "link". I would prefer another name, if we really need a concept separate from a button.
* The rendering to me is wrong. An `<a>` is a hyperlink, which this concept is not. By attaching an js-action to the `<a>` you in fact will cause confusion for every machine trying to semantically interpret the resulting html. AFAIK the new button tag was exactly introduced to solve this issue.

In general it seems to me as if this suggestion was build bottom up from the desired rendering result and not top down from the semantics we want to encode in the UI. It seem to confuse "renders as" and "is a".

Best regards!

**Comment by alex40724** [comment]
The important part: The links should be the single items in the drop down. For them we need to be able to use the JS/signaling processing. For each of them individually (some might just open an URL, other a modal).

I am also fine with passing an array of Buttons to the Dropdowns. I hope Timon can agree on this, otherwise I have no idea how to proceed, since he rejected DropdownItems and suggested Links.

Actually you did it yourself: https://github.com/ILIAS-eLearning/ILIAS/pull/544#discussion_r121111675

> So maybe a "link list" could be a node in the hierarchy, with the children "dropdown" and "breadcrumb".

But even if I pass an array of Buttons to the Dropdown I think we still need something that

> Links trigger interactions that change the view.

and

> Links MAY be used inline in a Textual Paragraph.

Would "Changing the view" (or screen?) be something that would fit the "references another resource in hypertext content" character?

And Buttons are clearly not allowed inline in Textual Paragraphs, right?

> It seem to confuse "renders as" and "is a".

I based the rendering part on your Button definition:

> Button is a clickable, graphically obtrusive control element.

Beside the inline use in text paragraphs this was the main differentiator to me. I think visual representation and "being something" should go hand in hand. Otherwise users might not learn what they get. And I barely think that they will identify the single items in a Dropdown as Buttons.

But as I said for me it is ok to shrink the link concept down to something that "changes views" and is allowed to be used inline in text paragraphs. Would this be ok?

And would open a modal be ok for a link (since it is some kind of other view/screen/resource)? Otherwise I do not know how to implement features like this: http://www.ilias.de/docu/goto.php?target=wiki_1357_Usability_Improvements_Marginal_List_Calendar

**Comment by Amstutz** [comment]
Hello @alex40724 @klees ,

I think we need the concept you are proposing here @alex40724. Thanks for this PR! 

I just had a talk with Richard and I think the issue he has boils down to the naming of this component. I somewhat feel responsible for this issue, since it was me initially proposed to name this "Link". However, Richards point that this is not really a proper Hyperlink if just a JS event is fired is valid. Alexandra and I somehow also stumbled over this while researching different button types last year. We stumbled upon this "Shy Button" or "btn-link" which is a button that mymics a link. 

Therefore I propose the following and hope that everybody can agree to that (if I understood @klees correctly in the previous talk, he seems fine with that):

- We implement this as Button -> Shy (Reads as "Shy Button")
- This is a button the mimics a link (including the a tag)

I think lots of the description/rules that you provided alex (big thanks for that) can be reused. There are some slight adjustements to be made.  If you want I can provide them.

**Comment by alex40724** [comment]
So we implement a Shy Button extending the Button and I can pass an array of Shy Buttons and Dividers to the Dropdown constructor, right?

And: If I want a "real hyperlink" in running text, I would not use a Shy Button, right?

**Comment by Amstutz** [comment]
@alex40724:

> So we implement a Shy Button extending the Button and I can pass an array of Shy Buttons and Dividers to the Dropdown constructor, right?

Yes to the Shy Buttons, yes to the dividers from my point of view. I did not talk the dividers through with @klees. See my comment in: https://github.com/ILIAS-eLearning/ILIAS/pull/550

> And: If I want a "real hyperlink" in running text, I would not use a Shy Button, right?

Regarding the description of "Buttons" no, it would be better to implement the Hyperlink for that. However, that is for another day.

**Comment by klees** [comment]
I'm very happy with the shy-button.

**Comment by alex40724** [comment]
Here we go with the Shy Button https://github.com/ILIAS-eLearning/ILIAS/pull/556

**Comment by alex40724** [comment]
@klees @Amstutz 

I have reduced to Link proposal here now to reflect the idea of a hypertext link (no JS anymore).

Feel free to suggest changes for description/rules and more in the comments.

File: src/UI/Factory.php [hunk 1]
```diff
@@ -336,4 +336,43 @@ public function panel();
...
+	 * description:
+	 *   purpose: >
+	 *      Links are used navigate to other resources or views of the system. Clicking
+	 *      on a link does not change the systems status.
+	 *   composition: >
+	 *      Link is a clickable, graphically minimal obtrusive control element. It can
+	 *      bear text or other content.
+	 *   effect: >
+	 *      On-click, the resource or view indicated by the link is requested and
+	 *      presented.
```
Comment (Amstutz) [inline src/UI/Factory.php]: Maybe add: "Links are not used to trigger JS-Events" or similar.

File: src/UI/Factory.php [hunk 2]
```diff
@@ -336,4 +336,43 @@ public function panel();
...
 
+	/**
+	 * ---
+	 * description:
+	 *   purpose: >
+	 *      Links are used navigate to other resources or views of the system. Clicking
+	 *      on a link does not change the systems status.
+	 *   composition: >
+	 *      Link is a clickable, graphically minimal obtrusive control element. It can
+	 *      bear text or other content.
```
Comment (Amstutz) [inline src/UI/Factory.php]: Maybe add: Links always contain a valid href tag (which should not just contain a hash tag).

File: src/UI/Factory.php (hunk 1 above)
Comment (Amstutz) [inline src/UI/Factory.php]: Maybe add: "Links are not used to trigger JS-Events" or similar.

File: src/UI/Factory.php [hunk 3]
```diff
@@ -336,4 +336,43 @@ public function panel();
...
+	 *   effect: >
+	 *      On-click, the resource or view indicated by the link is requested and
+	 *      presented.
+	 *   rivals:
+	 *      buttons: >
+	 *          Buttons are used to trigger Interactions that usually change the systems
+	 *          status. Buttons are much more obtrusive than links.
```
Comment (Amstutz) [inline src/UI/Factory.php]: Maybe add: Buttons might fire JS-Events, links not.

File: src/UI/Factory.php [hunk 4]
```diff
@@ -336,4 +336,43 @@ public function panel();
...
+	 *
+	 * rules:
+	 *   usage:
+	 *      1: >
+	 *           Links MAY be used inline in a Textual Paragraphs.
+	 *   interaction:
+	 *      1: >
+	 *           Hovering an active link should indicate a possible interaction.
```
Comment (Amstutz) [inline src/UI/Factory.php:363]: Links MUST not be used to fire JS Event. Use a Shy Button instead.

File: src/UI/Implementation/Component/Link/Link.php [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+
+/* Copyright (c) 2017 Alexander Killing <killing@leifos.de> Extended GPL, see docs/LICENSE */
+
+namespace ILIAS\UI\Implementation\Component\Link;
+
+use ILIAS\UI\Component as C;
+use ILIAS\UI\Implementation\Component\ComponentHelper;
+
+/**
+ * This implements commonalities between standard and primary buttons.
```
Comment (Amstutz) [inline src/UI/Implementation/Component/Link/Link.php]: Change this to link

**Review by Amstutz — COMMENTED** [review]
Thanks a lot for updating this. This is definitely JF ready. I have some suggestions on the description. Use them as you like.

Some comments:
- Depending on the outcome of the Data Type Link discussion, we should use this data type here. Since this is yet ongoing, string is fine with me for the moment.
- Add tests for the final merge.

**Review by klees — APPROVED** [review]

**Comment by klees** [comment]
Thanks a lot for your patience. If we came to an agreement about the "link"-type, I would add it here, but atm I do not think we still need it. @nhaagen will be able to reuse this in the meantime.

**Comment by matthiaskunkel** [comment]
Jour Fixe, June 19, 2017: Pull request accepted. Please consider Timon's suggestions in the discussion part.

**Comment by alex40724** [comment]
@Amstutz I modified the rules according to your suggestions and added tests. Can you approve this for the trunk?

**Review by Amstutz — APPROVED** [review]

**Pull request merged**